be/src/exec/pipeline/table_function_operator.cpp

```diff
@@ -202,20 +162,62 @@ ChunkPtr TableFunctionOperator::_build_chunk(const std::vector<ColumnPtr>& colum
-        SCOPED_TIMER(_table_function_exec_timer);
-        COUNTER_UPDATE(_table_function_exec_counter, 1);
-        _table_function_result = _table_function->process(_table_function_state, &_table_function_result_eos);
-        DCHECK_EQ(_input_chunk->num_rows() + 1, _table_function_result.second->size());
-    }
+Status TableFunctionOperator::_process_table_function() {
+    SCOPED_TIMER(_table_function_exec_timer);
+    COUNTER_UPDATE(_table_function_exec_counter, 1);
+    _input_index_of_first_result = _table_function_state->processed_rows();
+    _next_output_row = 0;
+    _next_output_row_offset = 0;
+
+    _table_function_result = _table_function->process(_table_function_state);
+    return _table_function_state->status();
 }
+
 Status TableFunctionOperator::reset_state(starrocks::RuntimeState* state, const std::vector<ChunkPtr>& refill_chunks) {
     _input_chunk.reset();
-    _remain_repeat_times = 0;
-    _input_chunk_index = 0;
-    _table_function_result_eos = false;
+    _input_index_of_first_result = 0;
+    _next_output_row_offset = 0;
+    _next_output_row = 0;
     _is_finished = false;
     return Status::OK();
 }
+
+void TableFunctionOperator::_copy_result(const std::vector<ColumnPtr>& columns, uint32_t max_output_size) {
+    DCHECK_LE(_next_output_row, _table_function_result.first.size());
+    DCHECK_LT(_next_output_row_offset, _table_function_result.second->size());
+    uint32_t curr_output_size = columns[0]->size();
+    const auto& fn_result_cols = _table_function_result.first;
+    const auto& offsets_col = _table_function_result.second;
+    while (curr_output_size < max_output_size && _next_output_row < fn_result_cols[0]->size()) {
+        uint32_t start = _next_output_row;
+        uint32_t end = offsets_col->get_data()[_next_output_row_offset + 1];
+        DCHECK_GE(start, offsets_col->get_data()[_next_output_row_offset]);
+        DCHECK_LE(start, end);
+        uint32_t copy_rows = std::min(end - start, max_output_size - curr_output_size);
+
+        // Build outer data, repeat multiple times
+        for (size_t i = 0; i < _outer_slots.size(); ++i) {
+            ColumnPtr& input_column_ptr = _input_chunk->get_column_by_slot_id(_outer_slots[i]);
+            Datum value = input_column_ptr->get(_input_index_of_first_result + _next_output_row_offset);
+            if (value.is_null()) {
+                DCHECK(columns[i]->is_nullable());
+                down_cast<NullableColumn*>(columns[i].get())->append_nulls(copy_rows);
+            } else {
+                columns[i]->append_value_multiple_times(&value, copy_rows);
+            }
+        }
+
+        // Build table function result
+        for (size_t i = 0; i < _fn_result_slots.size(); ++i) {
+            columns[_outer_slots.size() + i]->append(*(fn_result_cols[i]), start, copy_rows);
+        }
+
+        curr_output_size += copy_rows;
+        _next_output_row += copy_rows;
+        if (start + copy_rows == end) {
```

Comment:
better add `DCHECK_LE(start + copy_rows, end)` or change from `==` to `>=`, in case of somebody modifying the calculation of `copy_rows` by mistake later.

Reply:
Updated, PTAL

Reply:
IMHO, replace `==` with `>=` may make the potential issue you mentioned above more difficult to discover, so I choose to add a DCHECK.